// angular2-infinite-scroll/angular2-infinite-scroll.js
"use strict";
var infinite_scroll_1 = require('./src/infinite-scroll');
var scroller_1 = require('./src/scroller');
var position_resolver_1 = require('./src/position-resolver');
var axis_resolver_1 = require('./src/axis-resolver');
var infinite_scroll_2 = require('./src/infinite-scroll');
exports.InfiniteScroll = infinite_scroll_2.InfiniteScroll;
var scroller_2 = require('./src/scroller');
exports.Scroller = scroller_2.Scroller;
var position_resolver_2 = require('./src/position-resolver');
exports.PositionResolver = position_resolver_2.PositionResolver;
exports.PositionResolverFactory = position_resolver_2.PositionResolverFactory;
var axis_resolver_2 = require('./src/axis-resolver');
exports.AxisResolver = axis_resolver_2.AxisResolver;
exports.AxisResolverFactory = axis_resolver_2.AxisResolverFactory;
var index_1 = require('./src/index');
exports.InfiniteScrollModule = index_1.InfiniteScrollModule;
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = {
    directives: [infinite_scroll_1.InfiniteScroll, scroller_1.Scroller, axis_resolver_1.AxisResolver, position_resolver_1.PositionResolver]
};
